### ISSUE TITLE
Amica can't reach basic-openai-api-wrapper

### ISSUE THREAD

**Issue by gitwittidbit** [semperai/amica#96]
Hi,
I installed Amica via docker compose and got it working with OpenAI.
But I want to use local providers and so I am trying to connect Amica to my Ollama server. Unlike other Frontends I am running (e.g. LibreChat), Amica can't connect to Ollama with the same info - I keep getting a network error.
What am I missing?
Thanks!

**Comment by slowsynapse** [comment]
Not sure, by local providers you mean you are running Ollama with on your computer? I am not familiar with LibreChat, can you provide more info about the error you are getting?

**Comment by gitwittidbit** [comment]
Yes, I have Ollama running on my local server (but in a different docker stack). Another frontend that works with Ollama is Open-Webui.
So when I chat with Amica (and have Ollama configured as backend), I am getting this: 

![Screenshot 2024-04-25 at 11 00 02](https://github.com/semperai/amica/assets/83287661/1c3a43a0-9c3b-44ed-9374-3f040ca97648)

**Comment by slowsynapse** [comment]
Hey, can you grab a screenshot from developer console in browser? Wondering if CORS issue.

**Comment by gitwittidbit** [comment]
Okay, so it says that the loading of mixed content was blocked:

![Screenshot 2024-04-25 at 13 18 50](https://github.com/semperai/amica/assets/83287661/9fa45224-944e-4ae9-9b00-834ec9556291)

Which is strange to me, because the link is http.

**Comment by napiquet** [comment]
when you say "link is http", do you mean that Amica is running on HTTP?

Both Amica and Ollama must be either on HTTP or HTTPS. You cannot have Amica running HTTPS and try to connect to Ollama with classic HTTP.
You would either need to change Ollama to run securely as well, or downgrade Amica to non secure.

**Comment by gitwittidbit** [comment]
I meant that Ollama was running on HTTP.

But I was actually accessing Amica via HTTPS. This is the same setup I use with open-webui, LibreChat and a few others.

I now tried accessing Amica also via HTTP but again got that network error. But this time, the developer console actually mentions CORS (as @slowsynapse had suggested earlier):

![Screenshot 2024-04-25 at 14 57 49](https://github.com/semperai/amica/assets/83287661/b12e408a-82b7-4712-ba97-cd313a1f6121)

It says that the request was blocked because of the same-origin-rule (reason: CORS headline 'Access-Control-Allow-Origin' missing). Status code 403

**Comment by napiquet** [comment]
Oh, did you allow all origins (*) in Ollama? This is required when accessing from another container stack.

For Amica, you still need to be on HTTP (or https) for both, but you also need ollama to allow the remote origin.

**Comment by gitwittidbit** [comment]
Not sure where in Ollama I can set that.

But I can say that open-webui and LibreChat (and a few other frontends) all running on the same host (but in different docker stacks) all can access Ollama. So it would seem that Ollama isn't overly picky about the origin of the requests the way it is configured at the moment.

(And I didn't want to over complicate things and solve one issue at a time, so I didn't mention it before, but Amica also can't connect to the TTS/STT backend with the openai-wrapper as well that is running directly on the same host (not in docker)).

I'm happy to make any changes to my Ollama setup if that helps. I just don't know where...

**Comment by napiquet** [comment]
If your Amica container cannot access TTS/STT either, I believe the problem comes from your Amica installation, not Ollama. I also run openwebui and I know you wouldn't be able to use it if your Ollama install wasn't already allowing all origins.

Are you accessing the app through the local network and where your backends are?
I personally didn't manage to use Ollama when accessing Amica through a public domain, I had to access the container directly with its IP on my network.

**Comment by gitwittidbit** [comment]
I can access Amica via HTTPS which goes through a domain on the internet and I can access it via HTTP under its local IP address. The IP address is the same for the other frontends and the Ollama (and TTS/STT) backend. (But I myself am on a different local network.)

**Comment by gitwittidbit** [comment]
Tried it again today.

All external services seem to be reachable. But none of the local services can be accessed. How can that be?

**Comment by slowsynapse** [comment]
Not sure. I think would need your exact setup to see if the issue replicates on another environment.

**Comment by gitwittidbit** [comment]
I gave this another try today but to no avail.

So, my setup is: Amica is running in a Debian 12 VM with rootless docker that runs on a Proxmox host. Ollama is running also in a Debian 12 VM with rootless docker on a Proxmox host. I also have librechat, open-webui, sillytavern, dify, big-AGI and others in the same environment and with comparable setups and all of them are able to contact ollama locally (on http, despite being accessed via https themselves).

Based on your comment above that both Amica and Ollama need to be either both on http or both on https, I made Ollama available via an external https proxy, so that both are now accessible via https. But this does not make a difference. I am still getting the network error when prompting Amica.

When I switch the chatbot backend to ChatGPT, Amica can connect to it without an error. So what is the difference here?

Thanks!

**Comment by slowsynapse** [comment]
@gitwittidbit did you figure out the issue?

**Comment by gitwittidbit** [comment]
> @gitwittidbit did you figure out the issue?

Unfortunately I didn't.

**Comment by slowsynapse** [comment]
@gitwittidbit To clarify:

1) The networkerror is occurring on amica client running where? Is this a webclient connecting remote to your said server, or is this your own computer?

2) You are running this server as VPS, and its on one server/computer?

3) Which specific version of Ollama is this?

4) Is it still flagging it as a CORS issue (because if it isn't its some other problem)

**Comment by slowsynapse** [comment]
I'm really not sure how to recreate this issue, if its a CORS issue, is cross origin enabled on both sides to * ?
